### ISSUE TITLE
Discrepancy in the number of concepts learned by the BAMultiShapes model

### ISSUE THREAD

**Issue by Qianli-Wu** [steveazzolin/gnn_logic_global_expl#1]
Hello,

I am trying to reproduce the results of your recently published paper and I've noticed a discrepancy between the reported results and my observations.

The paper mentioned that the BAMultiShapes model learns five different concepts, but when I load the provided best model (`BAMultishapes.pt`) under the `trained_models` folder, the model could only produce one concept.

<img width="1138" alt="Screenshot 2023-07-10 at 14 46 52" src="https://github.com/steveazzolin/gnn_logic_global_expl/assets/97045514/0f6f6956-23b7-4659-be54-735bf76253d3">

> While the screenshot is from a Mac (where Python 3.7 is not available), the outcome on Linux is the same.

Also, I have trained another GLGExplainer model following the instructions provided in the README.md file. Here, I am able to identify six different concepts, but only five of these match the shapes described in the paper. Is this a common occurrence for GLGExplainer?

![image](https://github.com/steveazzolin/gnn_logic_global_expl/assets/97045514/d81eebbe-aa9c-4bfc-9274-4dbc72669994)
> graphs in P1 do not match, and there's only one concept of Grid (instead of 2)

Could you please clarify if there is a specific reason for this discrepancy, and if there are any changes I need to make to the architecture or hyperparameters to reproduce the results as reported in the paper?

I appreciate your help.

**Comment by steveazzolin** [comment]
Hello,

I just tried cloning the repo on my personal machine to check the results, and by running `BAMultiShapes.ipynb` I actually managed to produce the same output as the one already present in the notebook. 

Could you please make sure that the torch, pyg, and torch-explain versions match the ones used in the codebase?

> the outcome on Linux is the same

Do you mean is the same as the Mac screen or the same wrt my results?


> Could you please clarify if there is a specific reason for this discrepancy, and if there are any changes I need to make to the architecture or hyperparameters to reproduce the results as reported in the paper?


Since there is no explicit supervision on the concepts, it can be that the concepts learned are not the same and I would not be surprised if your clustering actually clusters together all instances of the grid. This is not a problem at all, provided the final clusters are homogeneous and the accuracy satisfactory.

**Comment by Qianli-Wu** [comment]
Thank you for your prompt response and clarifications! 

> Do you mean is the same as the Mac screen or the same wrt my results?

Sorry for any confusion. I was referring to the same outcome as depicted on the Mac screen.

As advised, I'm trying to match the versions of torch, pyg, and torch-explain used in your codebase. However, I couldn't find these specific versions in the repository. Could you please provide the exact versions of torch, pyg, and torch-explain you used for your experiments?

Also, in the env.yml file, `ananconda` is listed as a channel. I suspect this might be `anaconda`? Furthermore, `pyg` is listed as a dependency, but it seems it can only be installed from the `pyg` channel, which isn't listed.

In an effort to resolve these issues, I adjusted the `env.yml` to:
```
name: gnn
channels:
  - pyg
  - conda-forge
  - defaults
  - anaconda
```
However, this causes my conda process to terminate due to excessive memory or time consumption. 

Could you please review these changes and advise on any necessary amendments?

I appreciate your assistance in advance.

**Comment by steveazzolin** [comment]
> However, I couldn't find these specific versions in the repository.

You are indeed right, thanks for pointing out. The automatic tool provided by Conda for exporting envs doesn't seem to work properly.
I'll report here the specific version I used to reproduce the results:

python=3.7.1
pytorch=1.10.1
torch-explain=1.5.1
pyg=2.0.3


I'll try to export a self-contained `env.yml` for ease of usage, but in the meantime, I suggest you to manually install the required packages
